src/native/libs/System.Security.Cryptography.Native/pal_evp_pkey.c

```diff
@@ -580,3 +612,211 @@ EVP_PKEY* CryptoNative_LoadPublicKeyFromEngine(const char* engineName, const cha
+    if (store == NULL)
+    {
+        goto end;
+    }
+
+    // Quite similar to loading a single certificate from a PFX, if we find a private key that wins.
+    // Otherwise, the first public key wins.
+    // Otherwise, we'll push a keyload error
+    while (ret == NULL && !OSSL_STORE_eof(store))
+    {
+        OSSL_STORE_INFO* info = OSSL_STORE_load(store);
+
+        if (info == NULL)
+        {
+            continue;
+        }
+
+        int type = OSSL_STORE_INFO_get_type(info);
+
+        if (type == OSSL_STORE_INFO_PKEY)
+        {
+            ret = OSSL_STORE_INFO_get1_PKEY(info);
+            break;
+        }
+        else if (type == OSSL_STORE_INFO_PUBKEY && firstPubKey == NULL)
+        {
+            firstPubKey = info;
+            // skip the free
+            continue;
+        }
+
+        OSSL_STORE_INFO_free(info);
+    }
+
+    if (ret == NULL && firstPubKey != NULL)
+    {
+        ret = OSSL_STORE_INFO_get1_PUBKEY(firstPubKey);
+    }
+
+    if (ret == NULL)
+    {
+        ERR_clear_error();
+        ERR_put_error(ERR_LIB_NONE, 0, EVP_R_NO_KEY_SET, __FILE__, __LINE__);
+    }
+
+end:
+    if (firstPubKey != NULL)
+    {
+        OSSL_STORE_INFO_free(firstPubKey);
+    }
+
+    if (store != NULL)
+    {
+        OSSL_STORE_close(store);
+    }
+
+    if (ret == NULL)
+    {
+        if (prov != NULL)
+        {
+            assert(libCtx != NULL);
+            // we still want a separate check for libCtx as only prov could be NULL
+            OSSL_PROVIDER_unload(prov);
+        }
+
+        if (libCtx != NULL)
+        {
+            OSSL_LIB_CTX_free(libCtx);
+        }
+
+        *extraHandle = NULL;
+    }
+    else
+    {
+        EvpPKeyExtraHandle* extra = (EvpPKeyExtraHandle*)malloc(sizeof(EvpPKeyExtraHandle));
+        extra->prov = prov;
+        extra->libCtx = libCtx;
+        atomic_init(&extra->refCount, 1);
+        *extraHandle = extra;
+    }
+
+    return ret;
+#else
+    (void)providerName;
+    (void)keyUri;
+    ERR_put_error(ERR_LIB_NONE, 0, ERR_R_DISABLED, __FILE__, __LINE__);
+    *extraHandle = NULL;
+    return NULL;
+#endif
+}
+
+EVP_PKEY_CTX* CryptoNative_EvpPKeyCtxCreateFromPKey(EVP_PKEY* pkey, void* extraHandle)
+{
+    assert(pkey != NULL);
+
+#ifdef NEED_OPENSSL_3_0
+    if (API_EXISTS(EVP_PKEY_CTX_new_from_pkey))
+    {
+        EvpPKeyExtraHandle* handle = (EvpPKeyExtraHandle*)extraHandle;
+        OSSL_LIB_CTX* libCtx = (handle != NULL) ? handle->libCtx : NULL;
+        return EVP_PKEY_CTX_new_from_pkey(libCtx, pkey, NULL);
+    }
+    else
+#endif
+    {
+#ifndef NEED_OPENSSL_3_0
+        (void)extraHandle;
+#endif
+        return EVP_PKEY_CTX_new(pkey, NULL);
+    }
+}
+
+int32_t CryptoNative_EvpPKeyCtxConfigureForECDSASign(EVP_PKEY_CTX* ctx)
+{
+    if (ctx == NULL)
+    {
+        return 0;
+    }
+
+    if (EVP_PKEY_sign_init(ctx) <= 0)
+    {
+        return 0;
+    }
+
+    return 1;
+}
+
+int32_t CryptoNative_EvpPKeyCtxConfigureForECDSAVerify(EVP_PKEY_CTX* ctx)
```

Comment:
This doesn't seem like it does anything specific for ECDSA. It puts it in a verify init state. Could be used for other signature algorithms. Why don't we need to do this for RSA?

Reply:
this doesn't exist anymore since I matched the old style RsaSignHash